### DISCUSSION TITLE
Get system information in concurrency.

### DISCUSSION THREAD

**Discussion by youpong** [TanmayPatil105/procfetch#92] [Ideas]
This is an interesting technical theme, isn't it?

**Comment by TanmayPatil105** [comment]
Yess!
You must have realised how long time it takes to fetch the information where we have used `exec` because of pipe!
Like it takes `2s` to fetch package information.
Maybe we can `optimise` that!
I love `go` for this, how easy they have made `go routines` usable.

**Comment by youpong** [comment]
I have implemented [sleep sort](https://github.com/youpong/cpp-sleep-sort) on a trial basis. I think that the parallelization techniques used here can be used to parallelize this project.

**Comment by TanmayPatil105** [comment]
Will look into it 👍🏻!

**Comment by youpong** [comment]
In Node.js, we have asynchronous processing like this.
```javascript
const fs = require(‘fs’);

fs.readFile(‘README.md’, (err, data) => {
	if (err) throw err;
	console.log(data);
}
```

Following that, it would be nice to have asynchronous processing like this.(The code omits exclusive lock for pkgs.)
```c++
Command.exec(“dpkg -l”, [](auto c) {
	pkgs.push_back(rec{“dpkg”s, c.getOutputLines()});
});
```

**Reply by youpong** [reply]
The bellow function returns when the all threads execution have completed.
```c++
Command.wait()
```

**Reply by TanmayPatil105** [reply]
So, print `packages` after it returns.

**Comment by TanmayPatil105** [comment]
@youpong Checkout branch [threads](https://github.com/TanmayPatil105/procfetch/tree/threads)
Ignore the way code is written :)
Just see the logic behind

**Comment by youpong** [comment]
I think the program could be simplified by completely separating data collection and output.

For example, there is a while statement waiting for Info.theme to be retrieved in other thread. It will be no longer necessary to wait for each.

**Reply by TanmayPatil105** [reply]
But sequencing of the `cout` statements is important too right?
So we gotta print them sequentially and the while loops ensure it!!

**Reply by youpong** [reply]
It works correctly as it is now.

But it's not a good idea to use while as a way to wait for tasks to be done in other threads, as it uses CPU.
Waiting should be done with only join().

**Reply by TanmayPatil105** [reply]
Hmm I see

**Reply by youpong** [reply]
Outline:
1. multiple threads execute the task of storing the Info struct
2. wait for all threads to finish
3. cout the Info struct

**Reply by TanmayPatil105** [reply]
The only problem with this approach is that the `printing` stops after printing ascii art for some time and we don't want that?

**Reply by youpong** [reply]
As far as I can tell, it's not the ascii art display but getPackages() that's taking a long time.

**Reply by youpong** [reply]
If you comment out line 123, which is the call to getPackage(), it will run very quickly.

**Reply by youpong** [reply]
So it would be faster if getPackage() could be concurrent, I guess.
